core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -810,7 +810,7 @@ private[spark] object Utils extends Logging {
    * Get the local machine's URI.
    */
   def localHostNameForURI(): String = {
-    customHostname.getOrElse(InetAddresses.toUriString(localIpAddress))
+    sys.env.get("SPARK_PUBLIC_DNS").getOrElse( customHostname.getOrElse(InetAddresses.toUriString(localIpAddress)) )
```

Comment:
don't ask me about mesos internals; I work on YARN/hadoop stuff. I was just looking to see what was going to break there —and what are the usual general troublespots with apps (networking; configuration)